### PR TITLE
Added groups datasource support.

### PR DESCRIPTION
**Pull request by Turhsus**
Added groups datasource support and associated test cases. Added the group datasource to the provider. Refactored the GroupSchema resource to be public.

Adding documentation here in a second, but wanted the PR to go public so that we could start reviewing

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,10 +1,17 @@
+## 6.20.0 (December 19, 2022).
+
+IMPROVEMENTS:
+
+* Added new datasource: datasource/datasource_artifactory_group.go
+  PR: [#PR](https://github.com/jfrog/terraform-provider-artifactory/pull/607)
```
Comment (alexhung) [inline CHANGELOG.md:6]: Should include the original issue (https://github.com/jfrog/terraform-provider-artifactory/issues/548)

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,10 +1,17 @@
+## 6.20.0 (December 19, 2022).
```
Comment (alexhung) [inline CHANGELOG.md:1]: This version is out-of-dated. Please refresh your branch from `master`.

File: docs/data-sources/artifactory_group.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Artifactory File Info Data Source
+
+Provides an Artifactory group datasource. This can be used to read the configuration of groups in artifactory.
+
+## Example Usage
+
+```hcl
+#
+data "artifactory_group" "my_group" {
+			name  = "my_group"
```
Comment (alexhung) [inline docs/data-sources/artifactory_group.md:10]: Too many indentations.

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/jfrog/terraform-provider-artifactory/v6/pkg/artifactory/resource/security"
+	"github.com/jfrog/terraform-provider-shared/packer"
+	"github.com/jfrog/terraform-provider-shared/predicate"
+	"github.com/jfrog/terraform-provider-shared/util"
+	"golang.org/x/net/context"
+	"net/http"
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:13]: Please sort and group the imports per:

1) Golang builtins,
2) The others 😄

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		"include_users": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Default:     false,
+			Description: "Setting includeUsers to true will return the group with its associated user list attached.",
+		},
+	}
+
+	baseGroupSchema := security.GroupSchema
+	delete(baseGroupSchema, "detach_all_users")
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:35]: This will delete `detach_all_users` from the `security.GroupSchema` map since you've only make a reference, not a deep copy.

In this case, call `util.MergeMaps` first which deep copies, then delete the attribute.

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+func ArtifactoryGroupSchema() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: DataSourceGroupRead,
+		Schema:      getDataSourceGroupSchema(),
+		Description: "Provides the Artifactory Group data source. Contains information about the group configuration and optionally provides its associated user list.",
+	}
+}
+
+func getDataSourceGroupSchema() map[string]*schema.Schema {
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:24]: These funcs should be inside `ArtifactoryGroupSchema()` to limit their scope.

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	baseGroupSchema := security.GroupSchema
+	delete(baseGroupSchema, "detach_all_users")
+	return util.MergeMaps(baseGroupSchema, includeUsersSchema)
+}
+
+func DataSourceGroupRead(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	group := security.Group{}
+	name := d.Get("name").(string)
+	includeUsers := d.Get("include_users").(string)
+	url := fmt.Sprintf("%s%s?includeUsers=%s", security.GroupsEndpoint, name, includeUsers)
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:43]: Let's not do string interpolation. Let's use Resty's [SetPathParams](https://pkg.go.dev/github.com/go-resty/resty/v2#Request.SetPathParams)

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	resp, err := m.(*resty.Client).R().SetResult(&group).Get(url)
+
+	if err != nil {
+		if resp != nil && (resp.StatusCode() == http.StatusBadRequest || resp.StatusCode() == http.StatusNotFound) {
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:47]: Not sure this is necessary at all.

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/jfrog/terraform-provider-artifactory/v6/pkg/acctest"
+	"github.com/jfrog/terraform-provider-artifactory/v6/pkg/artifactory/resource/security"
+	"github.com/jfrog/terraform-provider-shared/test"
+	"github.com/jfrog/terraform-provider-shared/util"
+	"log"
+	"strconv"
+	"testing"
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:12]: Group and sort please.

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	log.Printf("Create group %s done.", group.Name)
+}
+
+func TestAccGroup_basic_datasource(t *testing.T) {
+	id, tempFqrn, groupName := test.MkNames("test-group-full", "artifactory_group")
+	temp := `
+		data "artifactory_group" "{{ .groupName }}" {
+			name  = "{{ .groupName }}"
+      include_users = true
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:44]: Not enough indentations

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+		},
+		CheckDestroy: func(state *terraform.State) error {
+			restyClient := acctest.GetTestResty(t)
+			restyClient.R().Delete(security.GroupsEndpoint + groupName)
+			return nil
+		},
+	})
+}
+
+func TestAccGroup_basic_datasource_includeusers_false(t *testing.T) {
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:86]: Since `include_users` is default to `false` I think this test should be the opposite and test for true. Then in previous test, the hcl should not have `include_users` attribute at all which will test both the default value *and* the response parsing.

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+		CheckDestroy: func(state *terraform.State) error {
+			restyClient := acctest.GetTestResty(t)
+			restyClient.R().Delete(security.GroupsEndpoint + groupName)
+			return nil
+		},
+	})
+}
+
+// Found group
+// Not found group
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:124]: Tidy up please.

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "name", groupName),
+					resource.TestCheckResourceAttr(fqrn, "description", description),
+					resource.TestCheckResourceAttr(fqrn, "users_names.#", "0"), //Include users set to false, so no users should be in this list.
+				),
+			},
+		},
+		CheckDestroy: func(state *terraform.State) error {
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:115]: Let's DRY this destroy function.

File: pkg/artifactory/provider/provider.go [hunk 14]
```diff
@@ -168,6 +168,7 @@ func Provider() *schema.Provider {
 			map[string]*schema.Resource{
 				"artifactory_file":     datasource.ArtifactoryFile(),
 				"artifactory_fileinfo": datasource.ArtifactoryFileInfo(),
+				"artifactory_group":    datasource.ArtifactoryGroupSchema(),
```
Comment (alexhung) [inline pkg/artifactory/provider/provider.go:171]: No need for `Schema` in the func name

File: docs/data-sources/artifactory_group.md [hunk 15]
```diff
@@ -0,0 +1,36 @@
+# Artifactory File Info Data Source
```
Comment (alexhung) [inline docs/data-sources/artifactory_group.md:1]: Group not 'File Info'

File: docs/data-sources/artifactory_group.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+The following arguments are supported:
+
+* `name` - (Required) Name of the group.
+* `include_users` - (Optional) Determines if the group's associated user list will return as an attribute. Default is false.
+
+## Attribute Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `name`                - Name of the group
```
Comment (alexhung) [inline docs/data-sources/artifactory_group.md:26]: Redundant.

**Review by alexhung — CHANGES_REQUESTED** [review]

File: pkg/artifactory/datasource/datasource_artifactory_group.go (hunk 6 above)
Comment (Turhsus) [reply pkg/artifactory/datasource/datasource_artifactory_group.go:24]: by these funcs you mean getDataSourceGroupSchema and DataSourceGroupRead?

File: pkg/artifactory/datasource/datasource_artifactory_group.go (hunk 8 above)
Comment (Turhsus) [reply pkg/artifactory/datasource/datasource_artifactory_group.go:47]: You don't think we need error handling if the user tries to get a group that doesn't exist?

File: pkg/artifactory/datasource/datasource_artifactory_group.go (hunk 6 above)
Comment (alexhung) [reply pkg/artifactory/datasource/datasource_artifactory_group.go:24]: Yep

File: pkg/artifactory/datasource/datasource_artifactory_group.go (hunk 8 above)
Comment (alexhung) [reply pkg/artifactory/datasource/datasource_artifactory_group.go:47]: We already return diag.FromErr() if there's any error from the request. This extra check is not necessary for data source. For resource, this extra code is to reset TF ID in the state.

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,3 +1,11 @@
+## 6.22.0 (December 19, 2022). Tested on Artifactory 7.47.12
+
+IMPROVEMENTS:
+
+* Added new datasource: datasource/datasource_artifactory_group.go
+  PR: [#PR](https://github.com/jfrog/terraform-provider-artifactory/pull/607)
+  Issue: [#548](https://github.com/jfrog/terraform-provider-artifactory/issues/548)
```
Comment (alexhung) [inline CHANGELOG.md:7]: Please follow the examples in this file, where `Issue` comes first.

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+		},
+	})
+}
+
+func TestAccGroup_basic_datasource_includeusers_true(t *testing.T) {
+	id, tempFqrn, groupName := test.MkNames("test-group-full", "artifactory_group")
+	temp := `
+		data "artifactory_group" "{{ .groupName }}" {
+      name  = "{{ .groupName }}"
+      include_users = true
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group_test.go:94]: Inconsistent indentations.

File: pkg/artifactory/resource/security/resource_artifactory_group.go [hunk 19]
```diff
@@ -151,6 +151,34 @@ func groupParams(s *schema.ResourceData) (Group, bool, error) {
 	return group, includeUsers, nil
 }
 
+func GetDataSourceGroupSchema() map[string]*schema.Schema {
```
Comment (alexhung) [inline pkg/artifactory/resource/security/resource_artifactory_group.go:154]: I just realized these changes are made to the Resource file, not the Data Source file. These should only be applied to the data source. There should be minimal changes to the resource.

File: pkg/artifactory/resource/security/resource_artifactory_group.go [hunk 20]
```diff
@@ -151,6 +151,34 @@ func groupParams(s *schema.ResourceData) (Group, bool, error) {
...
+	delete(includeGroupSchema, "detach_all_users")
+	return includeGroupSchema
+}
+
+func DataSourceGroupRead(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	group := Group{}
+	name := d.Get("name").(string)
+	includeUsers := d.Get("include_users").(string)
+	m.(*resty.Client).R().SetResult(&group).SetQueryParam("includeUsers", includeUsers).Get(GroupsEndpoint + name)
+
```
Comment (alexhung) [inline pkg/artifactory/resource/security/resource_artifactory_group.go:174]: Re my previous comment on removing the 404 check, etc. I didn't realize this is the resource. I thought this func is the copy in the data source. Please revert these changes here, and apply them to the data source instead.

**Review by alexhung — CHANGES_REQUESTED** [review]
Changes of schema and the read func should be applied to the data source func, *not* the resource.

File: CHANGELOG.md (hunk 17 above)
Comment (Turhsus) [reply CHANGELOG.md:7]: I did, the first example I could find is version 6.9.0 and 6.7.2 where PR comes first.

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,3 +1,11 @@
+## 6.22.0 (December 19, 2022). Tested on Artifactory 7.47.12
+
+IMPROVEMENTS:
+
+* Added new datasource: datasource/datasource_artifactory_group.go
+  Issue: [#548](https://github.com/jfrog/terraform-provider-artifactory/issues/548)
+  PR: [#PR](https://github.com/jfrog/terraform-provider-artifactory/pull/607)
```
Comment (alexhung) [inline CHANGELOG.md:7]: Missing PR number inside the link `[]`.

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"github.com/go-resty/resty/v2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/jfrog/terraform-provider-artifactory/v6/pkg/artifactory/resource/security"
+	"github.com/jfrog/terraform-provider-shared/packer"
+	"github.com/jfrog/terraform-provider-shared/predicate"
+	"github.com/jfrog/terraform-provider-shared/util"
+	"golang.org/x/net/context"
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:11]: Sort and group imports. @danielmkn can help you configure IntelliJ to sort these correctly.

File: pkg/artifactory/datasource/datasource_artifactory_group.go [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+		m.(*resty.Client).R().SetResult(&group).SetQueryParam("includeUsers", includeUsers).Get(security.GroupsEndpoint + name)
+
+		d.SetId(group.Name)
+
+		pkr := packer.Universal(predicate.SchemaHasKey(security.GroupSchema))
+		return diag.FromErr(pkr(&group, d))
+	}
+
+	getDataSourceGroupSchema := func() map[string]*schema.Schema {
+		var includeUsersSchema = map[string]*schema.Schema{
```
Comment (alexhung) [inline pkg/artifactory/datasource/datasource_artifactory_group.go:28]: Not a blocker. In future, let's be consistent with style. Either use `:=` or `var`, and not mix them. (Unless the compiler forces your to 😄 )

File: pkg/artifactory/datasource/datasource_artifactory_group_test.go (hunk 18 above)
Comment (alexhung) [reply pkg/artifactory/datasource/datasource_artifactory_group_test.go:94]: Still not fixed.

**Review by alexhung — CHANGES_REQUESTED** [review]
Just a few minor updates needed

**Review by alexhung — APPROVED** [review]
